Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,115 +60,135 @@ bool VirtIOConsole::handle_device_config_change()
 
 void VirtIOConsole::handle_queue_update(u16 queue_index)
 {
-    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update");
-    VERIFY(queue_index <= TRANSMITQ);
-    switch (queue_index) {
-    case RECEIVEQ: {
-        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-        auto& queue = get_queue(RECEIVEQ);
+    dbgln_if(VIRTIO_DEBUG, "VirtIOConsole: Handle queue update {}", queue_index);
+
+    if (queue_index == CONTROL_RECEIVEQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_receive_buffer->lock());
+        auto& queue = get_queue(CONTROL_RECEIVEQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
 
-        auto used_space = m_receive_buffer->reserve_space(used).value();
-        auto remaining_space = RINGBUFFER_SIZE - used;
-
-        // Our algorithm always has only one buffer in the queue.
-        size_t unused;
-        VERIFY(queue.pop_used_buffer_chain(unused).is_empty());
-        popped_chain.release_buffer_slots_to_queue();
+        while (!popped_chain.is_empty()) {
+            popped_chain.for_each([&](auto addr, auto) {
+                auto offset = addr.as_ptr() - m_control_receive_buffer->start_of_region().as_ptr();
+                ControlMessage* message = (ControlMessage*)m_control_receive_buffer->vaddr().offset(offset).as_ptr();
+                process_control_message(*message);
+            });
```

Comment:
this can be more nicely written as a `do-while` loop

Reply:
I might be missing something here, but I don't see how the do-while equivalent would be nicer.

Reply:
it just removes a duplicated `VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);` call.